### PR TITLE
Improve np.array implementation

### LINKED ISSUE #1241: Wrong use of transpose in np.array() in fortran
**Describe the bug**
Wrong behaviour when using steps in views with np.array() bug in fortran translation

**To Reproduce**
Provide code to reproduce the behavior:
```python
import numpy as np
if __name__ == '__main__':
    a = np.array([[ 1,  2,  4,  8, 16],
       [ 1,  2,  4,  8, 16],
       [ 1,  2,  4,  8, 16],
       [ 1,  2,  4,  8, 16],
       [ 1,  2,  4,  8, 16]], order="F")
    tmp = a[::2]
    b = np.array(tmp, order="F")
    print(b)
```

Provide the generated code, or the error message:
```c
program prog_prog_test

  use test_0001

  use, intrinsic :: ISO_C_Binding, only : i64 => C_INT64_T
  use, intrinsic :: ISO_FORTRAN_ENV, only : stdout => output_unit
  implicit none

  integer(i64), allocatable, target :: a(:,:)
  integer(i64), pointer :: tmp(:,:)
  integer(i64), allocatable :: b(:,:)
  integer(i64) :: i
  integer(i64) :: i_0001
  integer(i64) :: i_0002

  allocate(a(0:4_i64, 0:4_i64))
  a = transpose(reshape([[1_i64, 2_i64, 4_i64, 8_i64, 16_i64], [1_i64, &
        2_i64, 4_i64, 8_i64, 16_i64], [1_i64, 2_i64, 4_i64, 8_i64, &
        16_i64], [1_i64, 2_i64, 4_i64, 8_i64, 16_i64], [1_i64, 2_i64, &
        4_i64, 8_i64, 16_i64]], [5_i64, 5_i64]))
  tmp(0:, 0:) => a(::2_i64, :)
  allocate(b(0:size(tmp, 1_i64, i64) - 1_i64, 0:size(tmp, 2_i64, i64) - &
        1_i64))
  b = transpose(tmp)
  write(stdout, '(A)', advance="no") '['
  do i = 0_i64, size(b, 1_i64, i64) - 1_i64 - 1_i64, 1_i64
    write(stdout, '(A)', advance="no") '['
    do i_0001 = 0_i64, size(b, 2_i64, i64) - 1_i64 - 1_i64, 1_i64
      write(stdout, '(I0 A)', advance="no") b(i, i_0001) , ' '
    end do
    write(stdout, '(I0 A)', advance="no") b(i, size(b, 2_i64, i64) - &
          1_i64) , ']'
    write(stdout, '()', advance="yes") 
  end do
  write(stdout, '(A)', advance="no") '['
  do i_0002 = 0_i64, size(b, 2_i64, i64) - 1_i64 - 1_i64, 1_i64
    write(stdout, '(I0 A)', advance="no") b(size(b, 1_i64, i64) - 1_i64, &
          i_0002) , ' '
  end do
  write(stdout, '(I0 A)', advance="no") b(size(b, 1_i64, i64) - 1_i64, &
        size(b, 2_i64, i64) - 1_i64) , ']'
  write(stdout, '(A)', advance="no") ']'
  write(stdout, '()', advance="yes") 
  if (allocated(a)) then
    deallocate(a)
  end if
  if (allocated(b)) then
    deallocate(b)
  end if

end program prog_prog_test
```

**Expected behavior**
Should print
```
[[1 2 4 8 16]
[1 2 4 8 16]
[1 2 4 8 16]]
```
Instead prints:
```
[[16 129 16]
[16 1 1]
[16 2 2]
[16 4 4]
[16 8 8]]
```

**Language**
fortran

 **Additional info**
Doesn't happen in 1D arrays

### PR DESCRIPTION
**Pull request by jalalium**
- Ensure `np.array` can be called with one variable argument (previously this caused an infinite loop over the elements of the array).
- Simplify `_print_NumpyArray` to reduce duplication and fix #1241 and fix duplicate #1257.
- Improve `np.array` implementation to correct the default order and add the `ndmin` parameter.
- Add tests for new code.
- Clean up `test_arrays.py` so failures reference the relevant issue, tuples use `==` instead of `np.array_equal`, and a more comprehensive check is carried out on returned arrays including checking the datatype and ordering.

### CODE REVIEW

**Comment by EmilyBourne** [comment]
@jalalium Thanks for your pull request. Please have a quick look at the [review process in the developer docs](https://github.com/pyccel/pyccel/blob/master/developer_docs/review_process.md) to see how we organise PRs in pyccel and add the appropriate labels. Thanks!

**Comment by yguclu** [comment]
@jalalium Please add a unit test which reproduces issue #1241. This should fail before your changes.

**Comment by jalalium** [comment]
> @jalalium Please add a unit test which reproduces issue #1241. This should fail before your changes.

While trying to add tests for both `epyccel` and `pyccel`, I noticed that their behavior are inconsistent. `Epyccel` performs an extra transposition of the `fortran` code. Basically, the output of the `epyccel` function and the output one would get from generating `fortran` code are inconsistent.
**For the following function:**
```    
def f():
    import numpy as np
    a = np.array([[1,  2,  4,  8, 16],
             [1,  2,  4,  8, 16],
             [1,  2,  4,  8, 16],
             [1,  2,  4,  8, 16],
             [1,  2,  4,  8, 16]], order="F")
    b = np.array(a, order="F")
    return b
```
**epyccel outputs the following**
```
f2 = epyccel(f, language='fortran')
print(f2())
```
```
[[ 1  1  1  1  1]
 [ 2  2  2  2  2]
 [ 4  4  4  4  4]
 [ 8  8  8  8  8]
 [16 16 16 16 16]]
 ```
 While the code generated in `Fortran` outputs the expected result.

**Comment by EmilyBourne** [comment]
I think there are 2 bugs here. I am opening issues

**Comment by EmilyBourne** [comment]
Unfortunately after looking into this I don't think you can add tests until #1128 and #1260 are merged. I am therefore marking this PR as blocked. Hopefully @ohachim can get back to PR #1128 soon so we can unblock these issues.

**Comment by EmilyBourne** [comment]
Blocked by #1258

File: pyccel/codegen/printing/fcode.py
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:1064]: I think there are other places where this function behaves badly. Eg:
```
a = np.ones((3,4), order='C')
b = np.ones((3,4), order='F')
a[:] = b
```

**Comment by EmilyBourne** [comment]
> It appears that issue #1496 has been fixed by PR #1477. If that is the case, I suppose that the conflicts in `pyccel/codegen/printing/cwrappercode.py` can be resolved by simply discarding any changes in this branch. @EmilyBourne Do you agree? 

That is correct. I needed a workaround to get the tests to pass but these changes can be discarded

File: pyccel/ast/numpyext.py
Comment (yguclu) [inline pyccel/ast/numpyext.py:581]: A more Pythonic one-liner:
```suggestion
                order = arg.order or 'C'
```

File: tests/epyccel/test_arrays.py
Comment (yguclu) [inline tests/epyccel/test_arrays.py:3865]: Can you please explain what goes wrong in the Python printer? A corresponding open issue should be referenced here.

File: tests/epyccel/test_arrays.py
Comment (yguclu) [inline tests/epyccel/test_arrays.py:3890]: Same comment as above.

File: tests/epyccel/test_arrays.py
Comment (yguclu) [inline tests/epyccel/test_arrays.py:3915]: Same comment as above.

File: tests/epyccel/test_arrays.py
Comment (yguclu) [inline tests/epyccel/test_arrays.py:3940]: Same comment as above.

File: tests/epyccel/test_epyccel_return_arrays.py
Comment (yguclu) [inline tests/epyccel/test_epyccel_return_arrays.py:808]: #1260 was merged last week.

**Review by yguclu — CHANGES_REQUESTED** [review]
Looks good, but I have a couple of questions.

File: tests/epyccel/test_epyccel_return_arrays.py
Comment (EmilyBourne) [inline tests/epyccel/test_epyccel_return_arrays.py:810]:
```suggestion
```

File: tests/epyccel/test_arrays.py
Comment (EmilyBourne) [reply tests/epyccel/test_arrays.py:3865]: The problem is that the code is different for the different orders. This is because the order is printed explicitly. This could be avoided by saving the initial order as we now do for the dtype. But I'm not convinced it is worth it as this use case is not common.

There is no existing relevant issue. There was one but it was converted to discussion #1334. This could be referenced but I am not sure that it helps clarify the reason for the xfail. Additionally when a fix is chosen and implemented it should fix all Python xfail marks so referencing the discussion shouldn't make any difference for finding xfails to remove

File: tests/epyccel/test_epyccel_return_arrays.py
Comment (EmilyBourne) [reply tests/epyccel/test_epyccel_return_arrays.py]: Removed

File: tests/epyccel/test_arrays.py
Comment (yguclu) [reply tests/epyccel/test_arrays.py:3865]: Could you please give me an example of how the code differs based on the order of the input array?

Comment (EmilyBourne) [reply tests/epyccel/test_arrays.py:3865]: Eg here:
```python
@template('T', ['int[:]', 'int[:,:]', 'int[:,:,:]', 'int[:,:](order=F)', 'int[:,:,:](order=F)']) 
 def array_ndmin_1(x : 'T'): 
     from numpy import array 
     y = array(x, ndmin=1) 
     return y
```
The argument types `int[:,:]`, and `int[:,:](order=F)` don't work at the same time as the code generated for the first type is something like:
```python
 def array_ndmin_1(x : 'int[:,:]'): 
     from numpy import array 
     y = array(x, dtype=x.dtype, order='C') 
     return y
```
(You may have to double check the actual generated code). NumPy allows properties like the order to be changed when copying data

Comment (yguclu) [reply tests/epyccel/test_arrays.py:3865]: I see. The problem is that we compute the value of `order` explicitly, instead of keeping the original argument `order='K'`, which is indeed the default value for `numpy.array`.

Comment (EmilyBourne) [reply tests/epyccel/test_arrays.py:3865]: Yes. This makes our life a lot easier when working out the order of any results for printing in C/Fortran, but in cases like this it causes differences in printing

**Review by yguclu — APPROVED** [review]

**Pull request merged**
